### PR TITLE
Avoid over-constrained delayed-use register case

### LINKED ISSUE #11141: [Arm64/Unix] Crossgen triggering LSRA assert 'farthestRefPhysRegRecord != nullptr'
On version c111a57, It is affecting these files during crossgen

```
Microsoft.VisualBasic.dll
System.Private.Xml.dll
```

System.Private.Xml.dll
```
Assertion failed 'farthestRefPhysRegRecord != nullptr' in 'System.Xml.Xsl.IlGen.XmlILOptimizerVisitor:FoldArithmetic(int,ref,ref):ref:this' (IL size 792)

    File: /home/vmjenkins/workspace/Dotnet/build_and_test/src/jit/lsra.cpp Line: 5928
    Image: /home/vmjenkins/workspace/Dotnet/build_and_test/bin/tests/pr1/Tests/coreoverlay/crossgen
```

Microsoft.VisualBasic.dll
```
Assert failure(PID 9690 [0x000025da], Thread: 9690 [0x25da]): Assertion failed 'farthestRefPhysRegRecord != nullptr' in 'Microsoft.VisualBasic.CompilerServices.Utils:CopyArray(ref,ref):ref' (IL size 232)

    File: /home/vmjenkins/workspace/Dotnet/build_and_test/src/jit/lsra.cpp Line: 5928
    Image: /home/vmjenkins/workspace/Dotnet/build_and_test/bin/tests/pr1/Tests/coreoverlay/crossgen
```

### PR DESCRIPTION
**Pull request by CarolEidt**
When a node has a delayed-use register (either due to an RMW def or due to an internalRegDelayFree),
if its result has a fixed-use, don't propagate the register constraint to the def, because there's
no associated fixed-reference to ensure that it will not be used by the delay-free register.

Fix #11141

### CODE REVIEW

**Comment by CarolEidt** [comment]
@dotnet/jit-contrib PTAL

File: src/jit/lsra.cpp [hunk 1]
```diff
@@ -737,15 +737,29 @@ void LinearScan::associateRefPosWithInterval(RefPosition* rp)
...
+                // incorrectly allocate that register.
+                // TODO-CQ: This means that we may often require a copy at the use of this node's result.
+                // This case could be moved to BuildRefPositionsForNode, at the point where the def RefPosition is
+                // created, causing a RefTypeFixedRef to be added at that location. This, however, results in
+                // more PhysReg RefPositions (a throughput impact), and a large number of diffs that require
+                // further analysis to determine benefit.
+                // See Issue #11274.
                 RefPosition* prevRefPosition = theInterval->recentRefPosition;
                 assert(prevRefPosition != nullptr && theInterval->firstRefPosition == prevRefPosition);
+                noway_assert(prevRefPosition->treeNode != nullptr);
```
Comment (pgavlin) [inline src/jit/lsra.cpp]: Are we likely to avoid the problematic case if this `noway` fires?

Comment (CarolEidt) [inline src/jit/lsra.cpp]: I had at first thought that we would, but as this is a tree temp interval, it is likely to occur with MinOpts as well (if it occurs at all). But this is a fundamental requirement of tree temp intervals (that the def have a treeNode), so I will simply assert it and add this to the condition below, which will be conservative.

**Comment by CarolEidt** [comment]
@pgavlin @dotnet/jit-contrib ping.

**Comment by pgavlin** [comment]
Any asm diffs?

**Comment by CarolEidt** [comment]
@pgavlin - no desktop asm diffs; am currently have some problems with jit-diffs.
@sdmaclea this fixes #11141 but I will probably wait to merge this until after 2.0.

**Comment by pgavlin** [comment]
@CarolEidt I will try to run jit-diffs.

It looks like the only place this will affect us on xarch is when dealing with certain SIMD nodes (and then I think only on x86). If there are no diffs, I am pretty comfortable taking this for 2.0.

**Comment by CarolEidt** [comment]
@pgavlin @RussKeldorph I was finally able to run jit-diffs, and no diffs were reported. As @pgavlin points out, on x64/x86 the `isInternalRegDelayFree` attribute is used only for SIMD (both x64 and x86), though the over-constrained case cannot occur there are there are no SIMD consumers of xmm regs that have a fixed reg requirement.

**Comment by BruceForstall** [comment]
Note that jit-diffs works via crossgen, so there will be no SIMD code. That's an unfortunate limitation.

**Comment by CarolEidt** [comment]
@BruceForstall - Yes, I hadn't thought of that. However, I also ran desktop diffs, which showed no diffs.

**Comment by sdmaclea** [comment]
@CarolEidt Thanks

**Pull request merged**